### DISCUSSION TITLE
What happened between 2.8.1 and 2.9.0 of xUnit?

### DISCUSSION THREAD

**Discussion by gimlichael** [xunit/xunit#2994] [Question] [answered]
Hello,

I am currently maturing my code for a generic status monitor - and while doing some bump of libraries, I noticed my functional unit test would stop working when updating xUnit to the latest and greatest version.

Since the update is not major, I was surprised to learn that my `StatusControllerTest` goes into an infinite mode when running past `Post_ShouldCreateNewStatus`.

Nothing happens when running `Get_ShouldGetStatus`.

All steps works flawlessly with xUnit 2.8.1 - so, what happened in between?

I am asking because I would like to use latest-and-greatest of your excellent unit test library - but I am unable to figure out what is wrong.

You can find my repo here: https://github.com/codebeltnet/status-monitor

Running the `StatusControllerTest` should produce result like this:

![image](https://github.com/user-attachments/assets/62a5f545-e036-4b40-9951-3f7166f401ef)

However, if you change https://github.com/codebeltnet/status-monitor/blob/main/test/Codebelt.StatusMonitor.FunctionalTests/Codebelt.StatusMonitor.FunctionalTests.csproj to use 2.9.0 of `xunit` and  `xunit.runner.console`, then you might be able to give me a hint?

![image](https://github.com/user-attachments/assets/42e2fce8-ec8f-4a5d-b090-c1805c4f1db5)

In advance thanks - and keep up the otherwise excellent work.

**Comment by bradwilson** [comment]
In answer to your post's question: the release notes tell you what happened between 2.8.1 and 2.9.0: https://xunit.net/releases/v2/2.9.0

> Since the update is not major

How did you make that determination without reading the release notes?

> All steps works flawlessly with xUnit 2.8.1

That's not the case on my machine.

I have a hung test, and it may be for the same or a different reason, but it definitely appears like it's deadlocked Tasks. I saw that `QueueWorkerTest.ExecuteAsync_ShouldCreateAcceptedStatus` was hung, so I tried running just that test in a debugger, and it was still hung.

Strangely enough, I set a breakpoint on the first line of code and it was never reached. So I broke the debugger and saw this in the list of tasks:

![image](https://github.com/user-attachments/assets/f1ab6882-0403-404b-93e0-bd2ea216d9ad)

I would categorize the waiting tasks in three categories:

1. xUnit.net trying to run this test
2. Amazon doing initialization
3. System.Net.Http trying to make a connection

These all start here:

![image](https://github.com/user-attachments/assets/6923aacf-3c53-43b2-8bcc-34365e5b486d)

It appears as though the hang is AWS libraries trying to talk to something that isn't available.

Is it possible that you forgot to give me some instructions on how to successfully run these tests, other than just running them in Test Explorer?

**Reply by bradwilson** [reply, marked as answer]
Since your problem is in `StatusControllerTest` and not `QueueWorkerTest`, I decided to only run the `StatusControllerTest` tests. In 2.8.1, they all passed, but in 2.9.0, I could reproduce the hang in `Get_ShouldGetStatus`.

When running under the debugger, I noticed this in the Output window:

```
Exception thrown: 'System.InvalidOperationException' in xunit.execution.dotnet.dll
There is no currently active test.
```

This is something writing to `ITestOutputHelper` outside the context of a running test. In this case, the stack trace points at `Cuemon.Extensions.Xunit.Hosting.XunitTestLogger.Log`. My guess is that this class is created once and shared among many tests; the reason it passes the first test in the list is because it's using the `ITestOutputHelper` from that test. When the second test tries to run, it is trying to log something to the `ITestOutputHelper` instance from the first test, which is no longer valid. That exception is being thrown on a background thread that nobody is listening to, so it appears like the test is hung.

It makes sense why this surfaces in 2.9.0, because this is in the 2.9.0 release notes:

![image](https://github.com/user-attachments/assets/0079eae2-38b1-4549-bfc8-2920034342c4)

So the bug is in the lifetime of `XunitTestLogger` (it needs to be re-created for each test with the correct `ITestOutputHelper` instance for that test).

**Reply by gimlichael** [reply]
Hi Brad,

Thank you for taking the time to provide me a better understanding on what has happened. I will reply to your counter-questions seperately.

> In answer to your post's question: the release notes tell you what happened between 2.8.1 and 2.9.0: https://xunit.net/releases/v2/2.9.0
> 
> > Since the update is not major
> 
> How did you make that determination without reading the release notes?

Well, I did read the release notes - this was my first priority. My conclusion for non-major is merely in the SemVer convention; you bumped minor to 9 and kept major as-is. Maybe that was a wrong assumption from my side.

> 
> > All steps works flawlessly with xUnit 2.8.1
> 
> That's not the case on my machine.

Sorry, maybe I was not clear enough. The area of interest is purely the StatusControllerTest - not the QueueWorkerTest, as this works as expected (but does require to do a `docker compose up`).

**Reply by gimlichael** [reply]
> Since your problem is in `StatusControllerTest` and not `QueueWorkerTest`, I decided to only run the `StatusControllerTest` tests. In 2.8.1, they all passed, but in 2.9.0, I could reproduce the hang in `Get_ShouldGetStatus`.
> 
> When running under the debugger, I noticed this in the Output window:
> 
> ```
> Exception thrown: 'System.InvalidOperationException' in xunit.execution.dotnet.dll
> There is no currently active test.
> ```
> 
> This is something writing to `ITestOutputHelper` outside the context of a running test. In this case, the stack trace points at `Cuemon.Extensions.Xunit.Hosting.XunitTestLogger.Log`. My guess is that this class is created once and shared among many tests; the reason it passes the first test in the list is because it's using the `ITestOutputHelper` from that test. When the second test tries to run, it is trying to log something to the `ITestOutputHelper` instance from the first test, which is no longer valid. That exception is being thrown on a background thread that nobody is listening to, so it appears like the test is hung.

Thanks for the intel - I will for sure pursue this.

> 
> It makes sense why this surfaces in 2.9.0, because this is in the 2.9.0 release notes:
> 
> ![image](https://private-user-images.githubusercontent.com/16944/354943200-0079eae2-38b1-4549-bfc8-2920034342c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NDI4NzEsIm5iZiI6MTcyMjg0MjU3MSwicGF0aCI6Ii8xNjk0NC8zNTQ5NDMyMDAtMDA3OWVhZTItMzhiMS00NTQ5LWJmYzgtMjkyMDAzNDM0MmM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDA3MjI1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkZDNiMDY5YWEzMGFiZGZhNDQ0YzY5ODQ0NDA0NzAyZjVhMGU4OWQ5YTU3MTA1ZTc2ODE1MWE1OThmNTRmNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sYc_0XThDcAsuhet8zRbrC-Q_WtT94nMys0uRVw7gcQ)
> 

Thanks, and yes, I did suspect that it was somehow related to `ITestOutputHelper`, but since the majority of my projects that received this update did not have an impact, I was doubting myself if that was a right suspicion (and also, I had a hard time imagining a "bug" that triggers a hang on not an Exception providing guidance on to how I have failed my call).

With your insights, I will defiantly investigate further and share the solution in case others have experiences a similar issue.

> So the bug is in the lifetime of `XunitTestLogger` (it needs to be re-created for each test with the correct `ITestOutputHelper` instance for that test).

Again, thank you for taking the time to provide a thorough walk-through on how I may be able to improve my code to be compliant with your recent changes - much appreciated <3

**Reply by bradwilson** [reply]
> My conclusion for non-major is merely in the SemVer convention

xUnit.net v2 does not follow SemVer. It's versioning system has been in place since before SemVer existed.

We will follow SemVer with v3, because we've moved the "product version" into the package names (i.e., `xunit.v3` instead of `xunit`) so that we can properly rev the major version of the package without implying anything about the _product version_.

Hope that helps, and good luck!

**Comment by gimlichael** [comment]
For those interested - and thanks to @bradwilson insights - the solution was to move this code:

<img width="475" alt="image" src="https://github.com/user-attachments/assets/d168eb64-df06-4e45-af48-b9fad819d649">

To the constructor of the StatusControllerTest class:

<img width="671" alt="image" src="https://github.com/user-attachments/assets/2f95c5e9-b03a-4b5b-bb34-d3dd70a18c0f">

Now it works as expected - and the "code-smell" is minor (though I would have loved the logger to stay singleton like remainder of the loggers provided out-of-the-box in ASP.NET).

https://github.com/codebeltnet/status-monitor/commit/57fe934636325c0d083ecdd3f07d58d6ddd60e19

Cheers - and code with passion 👨‍💻️🎉

**Reply by gimlichael** [reply]
Actually first fixed with the forthcoming v9.0.0 release - details here: https://github.com/gimlichael/Cuemon/commit/cf9609de615be311685c91a33abfe8ae84d869b8
